# HG changeset patch
# User Yury Ivanovich <Yury.Ivanovich@linuxmail.org>
# Date 1763239854 -3600
# Node ID 75d654349f9f110fbb9b8d0610994c590de7fde1
# Parent  46333e9619307ae12cd71cd6d4a738d48a931765
Bug 1999822 - Write correct UTF-8 for search term values. r=mkmelin

nsIMsgSearchTerm.termAsString is used in MsgTermListToString() to write the search term values to a string that is later written to a file.

Differential Revision: https://phabricator.services.mozilla.com/D272343

diff --git a/mailnews/base/src/nsMsgAccountManager.cpp b/mailnews/base/src/nsMsgAccountManager.cpp
--- a/mailnews/base/src/nsMsgAccountManager.cpp
+++ b/mailnews/base/src/nsMsgAccountManager.cpp
@@ -2491,9 +2491,9 @@ nsresult VirtualFolderChangeListener::In
     NS_ENSURE_SUCCESS(rv, rv);
     nsCOMPtr<nsIMsgFilter> tempFilter;
     filterList->CreateFilter(u"temp"_ns, getter_AddRefs(tempFilter));
     NS_ENSURE_SUCCESS(rv, rv);
-    filterList->ParseCondition(tempFilter, searchTermString.get());
+    filterList->ParseCondition(tempFilter, searchTermString);
     NS_ENSURE_SUCCESS(rv, rv);
     m_searchSession =
         do_CreateInstance("@mozilla.org/messenger/searchSession;1", &rv);
     NS_ENSURE_SUCCESS(rv, rv);
diff --git a/mailnews/search/public/nsIMsgFilterList.idl b/mailnews/search/public/nsIMsgFilterList.idl
--- a/mailnews/search/public/nsIMsgFilterList.idl
+++ b/mailnews/search/public/nsIMsgFilterList.idl
@@ -61,9 +61,9 @@ interface nsIMsgFilterList : nsISupports
     nsIMsgFilter createFilter(in AString name);
 
     void saveToFile(in nsIOutputStream stream);
 
-    void parseCondition(in nsIMsgFilter aFilter, in string condition);
+    void parseCondition(in nsIMsgFilter aFilter, in AUTF8String condition);
     // this is temporary so that we can save the filterlist to disk
     // without knowing where the filters were read from initially
     // (such as the filter list dialog)
     attribute nsIFile defaultFile;
diff --git a/mailnews/search/public/nsIMsgSearchTerm.idl b/mailnews/search/public/nsIMsgSearchTerm.idl
--- a/mailnews/search/public/nsIMsgSearchTerm.idl
+++ b/mailnews/search/public/nsIMsgSearchTerm.idl
@@ -121,10 +121,10 @@ interface nsIMsgSearchTerm : nsISupports
     boolean matchFolderFlag(in nsIMsgDBHdr msg);
 
     readonly attribute boolean matchAllBeforeDeciding;
 
-    readonly attribute ACString termAsString;
-    boolean matchKeyword(in ACString keyword); // used for tag searches
+    readonly attribute AUTF8String termAsString;
+    boolean matchKeyword(in AUTF8String keyword); // used for tag searches
     attribute boolean matchAll;
     /**
      * Does the message match the custom search term?
      *
diff --git a/mailnews/search/src/nsMsgFilterList.cpp b/mailnews/search/src/nsMsgFilterList.cpp
--- a/mailnews/search/src/nsMsgFilterList.cpp
+++ b/mailnews/search/src/nsMsgFilterList.cpp
@@ -682,9 +682,9 @@ nsresult nsMsgFilterList::LoadTextFilter
             nsAutoString unicodeStr;
             NS_CopyNativeToUnicode(value, unicodeStr);
             CopyUTF16toUTF8(unicodeStr, value);
           }
-          err = ParseCondition(m_curFilter, value.get());
+          err = ParseCondition(m_curFilter, value);
           if (err == NS_ERROR_INVALID_ARG)
             err = m_curFilter->SetUnparseable(true);
           NS_ENSURE_SUCCESS(err, err);
         }
@@ -717,15 +717,15 @@ nsresult nsMsgFilterList::LoadTextFilter
 // I guess interior quotes will need to be escaped - ("foo\")")
 // which will get written out as (\"foo\\")\") and read in as ("foo\")"
 // ALL means match all messages.
 NS_IMETHODIMP nsMsgFilterList::ParseCondition(nsIMsgFilter* aFilter,
-                                              const char* aCondition) {
+                                              const nsACString& aCondition) {
   NS_ENSURE_ARG_POINTER(aFilter);
 
   bool done = false;
   nsresult err = NS_OK;
-  const char* curPtr = aCondition;
-  if (!strcmp(aCondition, "ALL")) {
+  const char* curPtr = aCondition.BeginReading();
+  if (aCondition.EqualsLiteral("ALL")) {
     RefPtr<nsMsgSearchTerm> newTerm = new nsMsgSearchTerm;
     newTerm->m_matchAll = true;
     aFilter->AppendTerm(newTerm);
     return NS_OK;
